### DISCUSSION TITLE
Clip child control by name

### DISCUSSION THREAD

**Discussion by glen-nicol** [AvaloniaUI/Avalonia#9457] [Q&A]
My goal is to clip a hole in my control based on the location of a child control as it moves around and changes size. The code below seemed like the obvious way to do that. But instead it produces a completely white window. Why is that?

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaClipTest.MainWindow"
        Title="AvaloniaClipTest">
  <Border
    Background="Gray"
    Clip="{Binding #ClipSource.RenderedGeometry}">
    <Grid
      ColumnDefinitions="*,*,*"
      RowDefinitions="*,*,*">
      <!-- Expect this ellipse to actually be white (the window background) if the clipping works -->
      <Ellipse
        x:Name="ClipSource"
        Grid.Row="1"
        Grid.Column="1"
        Fill="Red" />
    </Grid>
  </Border>
</Window>
```

My instinct tells me the child control hasn't been created when the Clip is initially set and thus results in some default clip geometry that clips everything (rather than nothing?). But in the dev tools I can see the Clip property is set to the correct thing(image below). I did a quick search through the codebase and I didn't see the ClipProperty passed to AffectsRender anywhere, so I substituted the border in the XAML for a custom border (see code below) that does watch the Clip property and that didn't work either.

I've seen a few issues recently about clipping to rounded corners being affected by the new composition rendering in the preview. Is this related or something else

Attempted in 10.18, 11p1, and 11p4.

![image](https://user-images.githubusercontent.com/5103200/202330984-38d97dae-3827-47f1-ac2e-727297666dfb.png)

```
internal class ClipWatcher : Border
{
    static ClipWatcher()
    {
        AffectsRender<ClipWatcher>(ClipProperty);
    }
}
```

**Comment by maxkatz6** [comment]
I am not an expert with rendering code, but...

> The code below seemed like the obvious way to do that.

Not sure, maybe I misunderstood, but from your code I would expect white window and red ellipse. Because border is clipped by the ellipse geometry instead of making a "hole". Similarly, how it works in wpf - https://learn.microsoft.com/en-us/dotnet/api/system.windows.uielement.clip?view=windowsdesktop-7.0#examples

> But instead it produces a completely white window. Why is that?

Try to enable your IDE to stop the debug when any exception is thrown. Just in case if there is any in the render thread. When it happens, whole window might be white or transparent.

**Reply by glen-nicol** [reply]
The debugger window doesn't show any exceptions nor does it break when all CLR exceptions are configured for break on throw. Also this is a simplified repro of another project where The element I am trying to clip doesn't show but the rest of the UI does.

As to white hole vs red circle... yeah that might be the result too if it worked. Hopefully there is a way to invert! But baby steps.

**Comment by timunie** [comment]
I feel like opacity mask is what you are looking for intead of clipping. This can make any shape of hole into your control.

**Reply by glen-nicol** [reply]
I forgot about opacity mask, I will have to try that. Thanks for the reminder!

**Reply by glen-nicol** [reply]
same result when I do this:
```xaml
<Border.OpacityMask>
  <VisualBrush Visual="{Binding #ClipSource.RenderedGeometry}" />
</Border.OpacityMask>
```

or just #ClipSource

**Comment by glen-nicol** [comment]
What I posted previously was a MRE, but in my actual project I have done this with partial success.

```xaml
<Control.Clip>
  <CombinedGeometry GeometryCombineMode="Exclude">
    <CombinedGeometry.Geometry1>
      <RectangleGeometry
        Rect="0 0 100000 10000" />
    </CombinedGeometry.Geometry1>
    <CombinedGeometry.Geometry2>
      <EllipseGeometry
        Rect="90 50 850 520">
        <EllipseGeometry.Transform>
          <ScaleTransform
            ScaleX="{Binding #ScaleRoot.(vu:ScaleHelper.LocalScale)}"
            ScaleY="{Binding #ScaleRoot.(vu:ScaleHelper.LocalScale)}" />
        </EllipseGeometry.Transform>
      </EllipseGeometry>
    </CombinedGeometry.Geometry2>
  </CombinedGeometry>
</Control.Clip>
<local:ScalingCanvas
  x:Name="ScaleRoot"
  ...
```

This creates the hole correctly at scale 1 when it initially loads. But when the scale changes due to the window resizing, the EllipseGeometry doesn't react to the binding expression. I also added an event handler for the ClipProperty to see when that changes. I only get 1 event. So that tells me the bindings I have setup either aren't valid or maybe something about the ClipProperty makes it so the binding isn't watched? I don't see any binding errors in the output though.
